scylla-cdc-replicator/src/replicator_consumer.rs

```diff
@@ -525,52 +499,23 @@ impl ReplicatorConsumer {
         values_for_update: &mut [Option<&'a CqlValue>],
         timestamp: i64,
     ) -> anyhow::Result<()> {
-        let empty_udt = CqlValue::UserDefinedType {
-            keyspace: "".to_string(),
-            type_name: "".to_string(),
-            fields: vec![],
-        };
-        let value = data.get_value(column_name).as_ref().unwrap_or(&empty_udt);
-        // Order of values: ttl, added elements, pk condition values.
-
-        let values_for_update = &mut values_for_update.to_vec();
-        values_for_update[1] = Some(value);
-
-        self.precomputed_queries
-            .update_udt_elements(values_for_update, timestamp, column_name)
-            .await?;
-
-        self.delete_udt_elements(column_name, data, timestamp, value, values_for_update)
-            .await
-    }
-
-    async fn delete_udt_elements<'a>(
-        &mut self,
-        column_name: &str,
-        data: &'a CDCRow<'_>,
-        timestamp: i64,
-        value: &CqlValue,
-        values_for_update: &mut [Option<&'a CqlValue>],
-    ) -> anyhow::Result<()> {
-        let deleted_set = Vec::from(data.get_deleted_elements(column_name));
-
-        if !deleted_set.is_empty() {
-            let udt_fields = value.as_udt().unwrap();
-            let removed_fields = deleted_set
-                .iter()
-                .map(|deleted_idx| {
-                    let idx = deleted_idx.as_smallint().unwrap();
-                    let (udt_fields_by_id, _) = udt_fields.get(idx as usize).unwrap();
-                    format!("{}.{} = null", column_name, udt_fields_by_id)
-                })
-                .join(",");
-
-            let values_for_update = &mut values_for_update.to_vec();
-            values_for_update.remove(1);
+        let value = data.get_value(column_name).as_ref().unwrap();
```

Comment:
The `test_udt_fields_update` test fails when I add another column and try to update it.

```diff
@@ -848,7 +848,7 @@ mod tests {
             name: "TEST_UDT_ELEMENTS_UPDATE".to_string(),
             partition_key: vec![("pk", "int")],
             clustering_key: vec![("ck", "int")],
-            other_columns: vec![("v", "ut")],
+            other_columns: vec![("v", "ut"), ("dummy", "int")],
         };
 
         let udt_schemas = vec![TestUDTSchema {
@@ -864,6 +864,7 @@ mod tests {
             "UPDATE TEST_UDT_ELEMENTS_UPDATE SET v.int_val = null, v.bool_val = false WHERE pk = 0 AND ck = 1",
             "UPDATE TEST_UDT_ELEMENTS_UPDATE SET v.int_val = 5 WHERE pk = 0 AND ck = 1",
             "UPDATE TEST_UDT_ELEMENTS_UPDATE SET v.int_val = null WHERE pk = 0 AND ck = 1",
+            "UPDATE TEST_UDT_ELEMENTS_UPDATE SET dummy = 1 WHERE pk = 0 AND ck = 1",
         ];
 
         test_replication_with_udt(schema, udt_schemas, operations)
```
```
failures:

---- replication_tests::tests::test_udt_fields_update stdout ----
unique_name: test_rust_1651469149_7
unique_name: test_rust_1651469150_39
thread 'replication_tests::tests::test_udt_fields_update' panicked at 'called `Option::unwrap()` on a `None` value', scylla-cdc-replicator/src/replicator_consumer.rs:502:58
```

`cdc$deleted_x == false || cdc$deleted_x == null` does not imply that `x` column is not-null. If the column `x` was unaffected in the original operation, then both `x` and `cdc$deleted_x` will be null/false - so you can't just unwrap here.